src/amp-story-embed.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** @enum {string} */
+const LoadStateClass = {
+  LOADING: 'loading',
+  LOADED: 'loaded',
+  ERROR: 'error',
+};
+
+/** @const {string} */
+const CSS = `
+  :host { all: initial; display: block; border-radius: 0 !important; width: 405px; height: 720px; overflow: auto; }
+  .story { height: 100%; width: 100%; flex: 0 0 100%; border: 0; opacity: 0; transition: opacity 500ms ease; }
+  main { display: flex; flex-direction: row; height: 100%; }
+  .loaded iframe { opacity: 1; }
+`;
+
+/**
+ * Note that this is a vanilla JavaScript class and should not depend on AMP
+ * services, as v0.js is not expected to be loaded in this context.
+ */
+export class AmpStoryEmbed {
+  /**
+   * @param {!Element} element
+   * @constructor
+   */
+  constructor(element) {
+    console.assert(element.childElementCount > 0, 'Missing configuration.');
+
+    /** @private {!Element} */
+    this.element_ = element;
+
+    /** @private {!Document} */
+    this.doc_ = element.ownerDocument;
+
+    /** @private {!Array<!HTMLAnchorElement>} */
+    this.stories_ = Array.prototype.slice.call(element.querySelectorAll('a'));
+
+    /** @private {?Element} */
+    this.rootEl_;
+
+    /** @private {?HTMLIframeElement} */
+    this.iframeEl_;
+  }
+
+  /** @public */
+  build() {
+    this.initializeShadowRoot_();
+
+    // TODO: Build all child iframes.
+    this.buildIframe_(0);
+  }
+
+  /** @private */
+  initializeShadowRoot_() {
+    // Create shadow root
+    const shadowRoot = this.element_.attachShadow({mode: 'open'});
+
+    // Inject default styles
+    const styleEl = this.doc_.createElement('style');
+    styleEl.textContent = CSS;
+    shadowRoot.appendChild(styleEl);
+
+    this.rootEl_ = this.doc_.createElement('main');
+    shadowRoot.appendChild(this.rootEl_);
+  }
+
+  /**
+   * @param {number} index
+   * @private
+   */
+  buildIframe_(index) {
+    const story = this.stories_[index];
+    this.iframeEl_ = this.doc_.createElement('iframe');
+    this.iframeEl_.setAttribute('src', story.href);
+    this.iframeEl_.style.backgroundImage = story.getAttribute('data-poster-portrait-src');
+    this.iframeEl_.classList.add('story');
+    this.initializeLoadingListeners_();
```

Comment:
You'd need to set the listeners before you set the source if you want error/load listeners to fire in all cases/browsers.

Reply:
Updated, but I wonder if it matters since the iframe is not in the DOM yet